File: LINKED LISTS/Insert Operations/InsertBefore.java
public class InsertBefore {

    // ==============================
    // NODE CLASS
    // ==============================

    static class Node {
        int data;
        Node next;
    }


    // ==============================
    // CREATE A NEW NODE
    // ==============================

    static Node createNode(int data) {
        Node newNode = new Node();
        newNode.data = data;
        newNode.next = null;
        return newNode;
    }


    // ==============================
    // TRAVERSE THE LIST
    // ==============================

    static void traverse(Node head) {
        Node current = head;

        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }

        System.out.println();
    }


    // ==============================
    // INSERT BEFORE A TARGET
    // ==============================

    static Node insertBefore(Node head, int target, int data) {

        /*
            This is the most complex of the four insert operations
            because we need to know the node BEFORE the target.

            A linked list node only has a .next pointer — it knows
            what comes after it, but NOT what came before it.

            So if we walk forward and land ON the target, it is
            already too late. We have no way to reach the previous node.

            The solution: use TWO pointers moving together.

                previous → trails one step behind current
                current  → the node we are currently examining

            Visual as the two pointers walk the list:

            Step 0 (start):
                previous = null
                current  = [10] → [20] → [30]

            Step 1 (current.data != target, so advance):
                previous = [10]
                current  = [20] → [30]

            Step 2 (current.data == target, stop!):
                previous = [10]
                current  = [20] → [30]

            Now we have exactly what we need:
                previous points to the node BEFORE the target.
                current  points to the target itself.
        */

        Node current = head;
        Node previous = null;    // null means "there is no previous node yet"

        // Walk forward until we find the target or run out of nodes
        while (current != null && current.data != target) {
            previous = current;        // remember where we just were
            current = current.next;    // move forward
        }

        // If current is null, we walked off the end without finding target
        if (current == null) {
            System.out.println("Target " + target + " was not found.");
            return head;
        }

        // We found the target. Create the new node.
        Node newNode = createNode(data);

        /*
            Now there are two sub-cases depending on where the target is.

            ─────────────────────────────────────────────────────────
            SUB-CASE A: target is the FIRST node  (previous == null)
            ─────────────────────────────────────────────────────────

            previous is still null, which means current IS head.
            There is no node before the target, so we use
            the exact same logic as insertAtBeginning:

                newNode.next = head;
                head = newNode;

            Before:   head
                       ↓
                      [10] → [20] → [30]

            After:    head
                       ↓
                      [5] → [10] → [20] → [30]


            ─────────────────────────────────────────────────────────
            SUB-CASE B: target is somewhere AFTER the first node
            ─────────────────────────────────────────────────────────

            previous → [10]
            current  → [20] → [30]

            We want to insert [15] between them.

            STEP 1: Connect the new node to current (the target):

                newNode.next = current;

            State:
                previous → [10]
                newNode  → [15] → [20] → [30]
                current  → [20] → [30]

            STEP 2: Connect previous to the new node:

                previous.next = newNode;

            State:
                [10] → [15] → [20] → [30]

            Why this order?
            If we did STEP 2 first (previous.next = newNode),
            we would lose the reference to [20]. We'd have no
            way to set newNode.next correctly.

            Always wire the NEW node's outgoing link first,
            then redirect the existing node to point to it.
        */

        if (previous == null) {
            // SUB-CASE A: target is the first node
            newNode.next = head;
            head = newNode;
        } else {
            // SUB-CASE B: target is somewhere in the middle or end
            newNode.next = current;      // STEP 1: new node points to target
            previous.next = newNode;     // STEP 2: previous now points to new node
        }

        return head;
    }


    // ==============================
    // MAIN
    // ==============================

    public static void main(String[] args) {

        /*
            Create the initial list:

            10 → 20 → 30 → null
        */

        Node head = createNode(10);
        Node node2 = createNode(20);
        Node node3 = createNode(30);

        head.next = node2;
        node2.next = node3;

        System.out.print("Original list:                    ");
        traverse(head);

        /*
            Insert 15 before 20.
            This tests the normal mid-list case.

            Before:   10 → 20 → 30
                            ↓
            After:    10 → 15 → 20 → 30
        */

        head = insertBefore(head, 20, 15);

        System.out.print("After inserting 15 before 20:     ");
        traverse(head);

        /*
            Insert 5 before 10.
            This tests SUB-CASE A — target is the first node.

            Before:   10 → 15 → 20 → 30
                       ↓
            After:    5 → 10 → 15 → 20 → 30
        */

        head = insertBefore(head, 10, 5);

        System.out.print("After inserting 5 before 10:      ");
        traverse(head);

        /*
            Try to insert before a value that does not exist.
            Should print a not-found message and leave the list unchanged.
        */

        head = insertBefore(head, 99, 88);

        System.out.print("After inserting 88 before 99:     ");
        traverse(head);
    }
}

File: LINKED LISTS/Insert Operations/insertBefore.c
#include <stdio.h>
#include <stdlib.h>


// Node structure
struct Node {
    int data;
    struct Node *next;
};


// Creates and initializes a new node
struct Node *createNode(int data) {

    struct Node *newNode = malloc(sizeof(struct Node));

    if (newNode == NULL) {
        return NULL;
    }

    newNode->data = data;
    newNode->next = NULL;

    return newNode;
}


// Traverses the list and prints every node
void traverse(struct Node *head) {

    struct Node *current = head;

    while (current != NULL) {
        printf("%d ", current->data);
        current = current->next;
    }

    printf("\n");
}


// Inserts a new node before the node containing target
struct Node *insertBefore(struct Node *head, int target, int data) {

    struct Node *current = head;
    struct Node *previous = NULL;

    /*
        We need two pointers while searching:

        previous → node before current
        current  → node we are currently examining

        Example:

        [10] → [20] → [30]
         ↑       ↑
      previous current


        When current reaches the target:

        previous → [10]
        current  → [20]

        To insert [15] between them:

        1. Make the new node point to current.
        2. Make previous point to the new node.

        Result:

        [10] → [15] → [20] → [30]


        Special case:

        If previous == NULL, current is the first node.

        There is no node before it, so we use the
        insert-at-beginning algorithm:

        newNode->next = head;
        head = newNode;
    */


    // Search for the target node
    while (current != NULL && current->data != target) {

        previous = current;
        current = current->next;
    }


    // Target was not found
    if (current == NULL) {

        printf("Target %d was not found.\n", target);

        return head;
    }


    // Create the new node
    struct Node *newNode = createNode(data);

    if (newNode == NULL) {

        printf("Memory allocation failed.\n");

        return head;
    }


    // Target is the first node
    if (previous == NULL) {

        newNode->next = head;
        head = newNode;
    }


    // Target is somewhere after the first node
    else {

        newNode->next = current;
        previous->next = newNode;
    }


    return head;
}


int main() {

    /*
        Create the initial list:

        10 → 20 → 30 → NULL
    */

    struct Node *head = createNode(10);

    struct Node *node2 = createNode(20);

    struct Node *node3 = createNode(30);


    // Connect the nodes
    head->next = node2;
    node2->next = node3;


    printf("Original list: ");
    traverse(head);


    /*
        Insert 15 before 20:

        Before:

        10 → 20 → 30

        After:

        10 → 15 → 20 → 30
    */

    head = insertBefore(head, 20, 15);


    printf("After inserting 15 before 20: ");
    traverse(head);


    /*
        Insert 5 before 10.

        This tests the special case where the target
        is the first node.

        Before:

        10 → 15 → 20 → 30

        After:

        5 → 10 → 15 → 20 → 30
    */

    head = insertBefore(head, 10, 5);


    printf("After inserting 5 before 10:  ");
    traverse(head);


    /*
        Try to insert before a value that doesn't exist.
    */

    head = insertBefore(head, 50, 40);


    /*
        Free all nodes.

        We traverse the list while freeing each node
        so that we don't leave dynamically allocated
        memory behind.
    */

    struct Node *current = head;

    while (current != NULL) {

        struct Node *temp = current;

        current = current->next;

        free(temp);
    }


    return 0;
}